blog/2025-11-03-strategies-for-large-images.md

```diff
@@ -0,0 +1,114 @@
+---
+title: K3s strategies for image consumption
+description: Master online and offline image loading techniques in K3s for ultra-fast application startup, even with multi-gigabyte containers.
+authors: manuelbuil
+hide_table_of_contents: true
+---
+
+Slow image pulls can be annoying and could increase Kubernetes startup times over a healthy threshold, particularly in resource-constrained or air-gapped environments. The situation gets worsened by new AI cloud native apps, which often rely on astronomically large images (several gigabytes). This post dives into K3s mechanisms, like pre-pulling images and embedded registry mirror, that can effectively improve the user's experience ensuring your cluster is ready the moment you need the images, especially in environments where network bandwidth might be constrained.
+
+
+## 📦 Online & Offline Strategies: The Power of Local Import ##
+
+K3s provides two core mechanisms for ensuring large images are available quickly, whether you are connected to an external registry (online) or deploying in an isolated environment (offline). The goal is to shift the time spent waiting on a slow network pull into a fast local load during K3s startup.
+
+1. Pre-Pulling Images via a Manifest File (Online)
+In scenarios with internet connectivity, the goal is to initiate image pulls as early and efficiently as possible. K3s can be instructed to sequentially pull a set of images into the embedded containerd store during startup. This is ideal for ensuring base images are ready the moment the cluster starts.
+
+Users can trigger a pull of images into the containerd image store by placing a simple text file containing the image names, one per line, in the /var/lib/rancher/k3s/agent/images directory. This can be done before K3s starts or while K3s is running.
+
+Imagine the file `example.txt` which contains:
+
+```yaml
+docker.io/pytorch/pytorch:2.9.0-cuda12.6-cudnn9-runtime
+```
+Before starting the k3s service in the node, do the following:
+
+```bash
+# 1. Create the images directory on the node
+mkdir -p /var/lib/rancher/k3s/agent/images
+
+# 2. Copy the manifest file (example.txt)
+cp example.txt /var/lib/rancher/k3s/agent/images
+```
```

Comment:
I don't understand why we're breaking this into so many steps, and copying the file around. Couldn't this be simplified to just:
`mkdir -p /var/lib/rancher/k3s/agent/images && echo docker.io/pytorch/pytorch:2.9.0-cuda12.6-cudnn9-runtime > /var/lib/rancher/k3s/agent/images/pytorch.txt`

Reply:
I think for these sorts of folks using an editor and copying files around is harder than just echoing content directly into the file from a shell.